bpf/lib/lb.h

```diff
@@ -1147,6 +1152,37 @@ lb4_select_backend_id(struct __ctx_buff *ctx __maybe_unused,
 # error "Invalid load balancer backend selection algorithm!"
 #endif /* LB_SELECTION */
 
+static __always_inline __u16
+lb4_select_backend_id(struct __ctx_buff *ctx,
+		      struct lb4_key *key,
+		      const struct ipv4_ct_tuple *tuple __maybe_unused,
+		      const struct lb4_service *svc)
+{
+	__u32 __maybe_unused lb_selection_rule = 0;
+	__u32 __maybe_unused slot = 0;
+	struct lb4_service __maybe_unused *be = NULL;
+
+#ifdef ENABLE_DSR_TUNL
+	if (lb4_svc_is_loadbalancer(svc) || lb4_svc_is_external_ip(svc)) {
+		lb_selection_rule = ctx_load_meta(ctx, CB_LB_SELECTION_RULE);
+
+		if (svc->local_count > 0) {
+			slot = (get_prandom_u32() % svc->local_count) + 1;
+			be = lb4_lookup_backend_slot(ctx, key, slot);
+			if (be) {
+				return be->backend_id;
+			}
+		}
+
+		if ((lb_selection_rule & LB_LOCAL_BACKEND_ONLY) == LB_LOCAL_BACKEND_ONLY) {
+			return 0;
+		}
+	}
```

Comment:
Ok, so this is basically for the case where we arrive at the backend, and pick a local backend based on the svc ip, therefore it has to be local. I think we could avoid all the special case code for service lookup if we had separate map entries that are marked strictly for local backends, so the lookup would have a bit in the key set that points to 'local' and there it only contains the local backends. Then we'd also don't introduce bias e.g. for random selection.